lib/promise_utils.js

```diff
@@ -36,9 +43,68 @@ function createDoneTask () {
   return task
 }
 
+/**
+ * Similar to the 'once' function from the 'events' module, but allows you to add a condition for when you want to
+ * actually handle the event, as well as a timeout. The listener is additionally removed if a timeout occurs, instead
+ * of with 'once' where a listener might stay forever if it never triggers.
+ * Note that timeout and checkCondition, both optional, are in the third parameter as an object.
+ * @param emitter - The event emitter to listen to
+ * @param event - The name of the event you want to listen for
+ * @param [timeout=0] - An amount, in milliseconds, for which to wait before considering the promise failed. <0 = none.
+ * @param [checkCondition] - A function which matches the same signature of an event emitter handler, and should return something truthy if you want the event to be handled. If this is not provided, all events are handled.
+ * @returns {Promise} A promise which will either resolve to an *array* of values in the handled event, or will reject on timeout if applicable. This may never resolve if no timeout is set and the event does not fire.
+ */
+function onceWithCleanup (emitter, event, { timeout = 0, checkCondition = undefined } = {}) {
+  const task = createTask()
+
+  const onEvent = (...data) => {
+    if (checkCondition && !checkCondition(...data)) {
+      return
+    }
+
+    task.finish(data)
+  }
+
+  emitter.addListener(event, onEvent)
+
+  if (timeout > 0) {
+    // For some reason, the call stack gets lost if we don't create the error outside of the .then call
+    const timeoutError = new Error(`Event ${event} did not fire within timeout of ${timeout}ms ${checkCondition && checkCondition.name}`)
+    sleep(timeout).then(() => {
+      if (!task.done) {
+        task.cancel(timeoutError)
+      }
+    })
+  }
+
+  task.promise.finally(() => emitter.removeListener(event, onEvent))
+
+  return task.promise
```

Comment:
Wouldn't that work?
```js
return Promise.race([
  task.promise,
  sleep(timeout).then(() => {
    if (!task.done) {
      task.cancel(timeoutError)
    }
          
    return task.promise
  })
]).finally(() => emitter.removeListener(event, onEvent))
```

Reply:
You may need to return task.promise from finally

Reply:
It might work, but that sleep.then is really hard to read IMO. I think the current code is fine, it's pretty similar

Reply:
The finally callback also doesn't take any params

Reply:
> is really hard to read IMO.

It's how you should use promises correctly

Reply:
If we'd return task in `task.cancel`, then it could be written as:

```
sleep(timeout).then(() => task.cancel(timeoutError).promise)
```

Reply:
> It's how you should use promises correctly

Promises don't expose done state, so I don't really think there is a "correct" way to do this.